File: engine_v2/turn.py
"""
Turn orchestration for engine_v2 - ported from engine/turn.py's run_turn
(not run_turn_and_log: no replay/visualization consumer for engine_v2
yet, same reasoning as battle.py skipping the structured per-round log).

Agents aren't a formal class hierarchy here (unlike v1's BaseAgent) - each
decision point takes a plain callback instead, looked up per-faction from
a {faction: callable} dict (mirroring v1's {faction: agent}). Deliberate:
the real "agent" for engine_v2 will eventually be a neural policy with
its own natural interface (masked observation in, action out) that
almost certainly won't look like BaseAgent's five methods - building that
class hierarchy now would likely just be thrown away later. Callbacks
compose fine with an eventual class-based wrapper (an agent object's
bound methods are themselves callables) without committing to that shape
now. Callback signatures:

  decide_buy(state, faction, legal_actions) -> [action, ...]
  decide_movement(state, faction, step, legal_mask) -> (hex_index, direction) or None
  decide_cavalry(state, faction, step, legal_mask) -> (hex_index, direction) or None
  decide_target(state, hex_index, faction) -> target_faction or None
  decide_rectification(state, hex_index, winner_faction) -> [{"origin_hex", "units"}, ...]
"""

import random

import numpy as np

from .battle import get_winner, rectify_overflow, resolve_full_battle
from .buy import apply_buy_phase, get_legal_buy_actions
from .income import apply_income_phase
from .movement import apply_movement_step, legal_cavalry_mask, legal_movement_mask
from .state import NO_FACTION, NO_ORIGIN, count_units_in_play
from .terrain import apply_terrain_effects

MOVEMENT_STEPS = 3
CAVALRY_STEPS = 4
DEFAULT_MAX_TURNS = 100


def _run_battle_phase(state, decide_target, decide_rectification, rng):
    """Resolves every pending battle, in state.battle_order (battle
    creation order - see state.py's module docstring for why this has to
    match v1's dict-insertion-order semantics rather than e.g. hex-index
    order: the dismount infantry cap tally below is shared across every
    battle resolved this turn, so processing order can affect outcomes
    near the cap)."""
    infantry_counts = {f: count_units_in_play(state, f, 0) for f in range(state.num_factions)}
    pending_hexes = list(state.battle_order)

    for hex_index in pending_hexes:
        if not state.locked[hex_index]:
            continue

        def target_fn(s, hidx, faction, _faction_agent=decide_target):
            return _faction_agent[faction](s, hidx, faction)

        resolve_full_battle(state, hex_index, target_fn, rng, infantry_counts)

        winner = get_winner(state, hex_index)
        if winner is None:
            state.army_faction[hex_index] = NO_FACTION
            state.army_units[hex_index] = 0
            state.locked[hex_index] = False
            state.battle_faction[hex_index] = NO_FACTION
            state.battle_origin[hex_index] = NO_ORIGIN
            state.battle_units[hex_index] = 0
            state.battle_round[hex_index] = 0
            state.battle_order.remove(hex_index)
        else:
            send_back = decide_rectification[winner](state, hex_index, winner)
            rectify_overflow(state, hex_index, winner, send_back)


def _update_elimination(state):
    for faction in range(state.num_factions):
        if not state.alive[faction]:
            continue
        has_city = bool(np.any(state.city_owner == faction))
        has_units = bool(np.any(state.army_faction == faction))
        if not has_city and not has_units:
            state.alive[faction] = False


def run_turn(state, decide_buy, decide_movement, decide_cavalry, decide_target, decide_rectification, rng=None):
    """Mutates and returns `state`. Each decide_* argument is a
    {faction: callable} dict - see module docstring for each callable's
    signature. Only alive factions are asked for decisions, same as v1."""
    rng = rng or random.Random()

    apply_income_phase(state)

    buy_actions = {}
    for faction in range(state.num_factions):
        if not state.alive[faction]:
            continue
        legal = get_legal_buy_actions(state, faction)
        buy_actions[faction] = decide_buy[faction](state, faction, legal)
    apply_buy_phase(state, buy_actions)

    for step in range(MOVEMENT_STEPS):
        actions = {}
        for faction in range(state.num_factions):
            if not state.alive[faction]:
                continue
            legal = legal_movement_mask(state, faction)
            actions[faction] = decide_movement[faction](state, faction, step, legal)
        apply_movement_step(state, actions, cavalry_only=False)

    for step in range(CAVALRY_STEPS):
        actions = {}
        for faction in range(state.num_factions):
            if not state.alive[faction]:
                continue
            legal = legal_cavalry_mask(state, faction)
            actions[faction] = decide_cavalry[faction](state, faction, step, legal)
        apply_movement_step(state, actions, cavalry_only=True)

    _run_battle_phase(state, decide_target, decide_rectification, rng)
    apply_terrain_effects(state)
    _update_elimination(state)

    state.turn_number += 1
    return state


def check_game_end(state, max_turns=DEFAULT_MAX_TURNS):
    if int(np.sum(state.alive)) <= 1:
        return True
    return state.turn_number >= max_turns


def tally_final_score(state):
    """3-category score, matching engine/turn.py's tally_final_score:
    cities (1 pt each) + military (top 3 remaining unit counts score
    3/2/1). Voting is stubbed in v1 too (0 contribution) - see that
    module's docstring."""
    scores = {f: 0 for f in range(state.num_factions)}

    for faction in range(state.num_factions):
        scores[faction] += int(np.sum(state.city_owner == faction))

    unit_counts = {
        faction: int(state.army_units[state.army_faction == faction].sum())
        for faction in range(state.num_factions)
    }
    ranked = sorted(unit_counts.items(), key=lambda kv: -kv[1])
    military_points = [3, 2, 1]
    for i, (faction, _count) in enumerate(ranked[:3]):
        scores[faction] += military_points[i]

    return scores

File: engine/src/turn.cpp
#include "oo/turn.hpp"

#include "oo/buy.hpp"
#include "oo/movement.hpp"
#include "oo/terrain.hpp"

namespace oo {

void run_battle_phase(GameState& state, const TurnDecisions& decisions, Rng& rng) {
    // One shared tally for the whole turn - see the header note on battle order.
    int32_t infantry_counts[MAX_FACTIONS];
    for (int f = 0; f < MAX_FACTIONS; ++f) {
        infantry_counts[f] = f < state.num_factions ? count_units_in_play(state, f, kInfantry) : 0;
    }

    // Snapshot of the pending list: resolving a battle mutates battle_order, and
    // reverting/rectifying can add or remove entries.
    int16_t pending[MAX_ACTIVE_BATTLES];
    const int n_pending = state.num_battles;
    for (int i = 0; i < n_pending; ++i) pending[i] = state.battles[i].hex;

    for (int i = 0; i < n_pending; ++i) {
        const int hex_index = pending[i];
        if (!state.locked(hex_index)) continue;

        resolve_full_battle(state, hex_index, decisions.target, decisions.ctx, rng,
                            infantry_counts);

        const int winner = get_winner(state, hex_index);
        if (winner < 0) {
            // Everyone wiped out simultaneously: the tile is left empty with no
            // one victorious.
            state.army_faction[hex_index] = NO_FACTION;
            for (int t = 0; t < NUM_UNIT_TYPES; ++t) state.army_units[hex_index][t] = 0;
            state.erase_battle(hex_index);
            continue;
        }

        const int owner = state.city_owner[hex_index];
        int cap = MAX_STACK_SIZE;
        if (owner != NO_FACTION && owner != winner) {
            if (state.is_capital[hex_index]) {
                cap = 0;  // uncapturable: the winner is evicted whole
            } else {
                state.city_owner[hex_index] = NO_FACTION;  // outpost destroyed, not captured
                state.victory_points[winner] += kOutpostDestroyVp;
            }
        }

        SendBack send_back;
        send_back.clear();
        decisions.rectification(state, hex_index, winner, cap, send_back, decisions.ctx);
        rectify_overflow(state, hex_index, winner, send_back, cap);
    }
}

void run_turn(GameState& state, const TurnDecisions& decisions, Rng& rng) {
    // --- Buy ---------------------------------------------------------------
    {
        static_assert(MAX_FACTIONS <= 10, "chosen[] is sized for the rulebook's 10-player max");
        ChosenBuyActions chosen[MAX_FACTIONS];
        LegalBuyActions legal;
        for (int f = 0; f < state.num_factions; ++f) {
            chosen[f].clear();
            get_legal_buy_actions(state, f, legal);
            decisions.buy(state, f, legal, chosen[f], decisions.ctx);
        }
        apply_buy_phase(state, chosen);
    }

    // --- Movement ----------------------------------------------------------
    for (int step = 0; step < kMovementSteps; ++step) {
        MoveActions actions;
        actions.clear();
        LegalMask legal;
        for (int f = 0; f < state.num_factions; ++f) {
            legal_movement_mask(state, f, legal);
            Move chosen{};
            if (decisions.movement(state, f, step, legal, chosen, decisions.ctx)) {
                actions.set(f, chosen.hex, chosen.dir);
            }
        }
        apply_movement_step(state, actions, rng, /*cavalry_only=*/false);
    }

    // --- Cavalry -----------------------------------------------------------
    for (int step = 0; step < kCavalrySteps; ++step) {
        MoveActions actions;
        actions.clear();
        LegalMask legal;
        for (int f = 0; f < state.num_factions; ++f) {
            legal_cavalry_mask(state, f, legal);
            Move chosen{};
            if (decisions.cavalry(state, f, step, legal, chosen, decisions.ctx)) {
                actions.set(f, chosen.hex, chosen.dir);
            }
        }
        apply_movement_step(state, actions, rng, /*cavalry_only=*/true);
    }

    run_battle_phase(state, decisions, rng);
    apply_terrain_effects(state);
    apply_collect_phase(state, decisions.resource_choice, decisions.ctx);

    state.turn_number += 1;
}

int get_game_winner(const GameState& state) {
    int top = state.victory_points[0];
    for (int f = 1; f < state.num_factions; ++f) {
        if (state.victory_points[f] > top) top = state.victory_points[f];
    }
    if (top < kVpToWin) return -1;

    int winner = -1;
    int best_settle = -1;
    int contenders = 0;
    for (int f = 0; f < state.num_factions; ++f) {
        if (state.victory_points[f] != top) continue;
        ++contenders;
        // Tie-break: whoever settled their capital LATER wins.
        if (winner < 0 || state.capital_settle_order[f] > best_settle) {
            winner = f;
            best_settle = state.capital_settle_order[f];
        }
    }
    (void)contenders;
    return winner;
}

bool check_game_end(const GameState& state, int max_turns) {
    if (get_game_winner(state) >= 0) return true;
    return max_turns >= 0 && state.turn_number >= max_turns;
}

}  // namespace oo
